### ISSUE TITLE
`library_private_types_in_public_api` should not be triggered on enum constructors

### ISSUE THREAD

**Issue by mernen** [dart-lang/linter#4470]
Example:

```dart
enum Foo {
  a(_Internal()),
  b.x(_Internal()),
  c._x(_Internal()),
  ;

  // Unnamed constructor:
  const Foo(_Internal details);
  //        ~~~~~~~~~ Invalid use of a private type in a public API.

  // Named (public) constructor:
  const Foo.x(_Internal details);
  //          ~~~~~~~~~ Invalid use of a private type in a public API.

  // Named (underscore) constructor:
  const Foo._x(_Internal details);
  // (no problems reported)
}

class _Internal {
  const _Internal();
}
```

I don’t think the lint should be applied here. Rationale: since enum constructors cannot be called outside the enum declaration, they should always be treated as if they were private, regardless of their name, and get a free pass to use private types.

Tested on DartPad — Dart SDK 3.0.5 and 3.1.0-171.0.dev, as well as earlier versions of Dart 2.x.

**Comment by pq** [comment]
At first blush, I'm inclined to agree here.  @bwilkerson: I'd be curious to get your thoughts?

**Comment by bwilkerson** [comment]
Enum constructors can't be invoked outside the enum, so it's my opinion that they should not be considered to be public API.

**Comment by lrhn** [comment]
And, inspired by another recent issue, neither can generative constructors of classes that are 
* `sealed`
* `abstract` and `interface`
* `abstract` and `final`.

In all three cases the class is abstract, so the constructor cannot be invoked or torn off,
and the class cannot be subclassed, so no the constructor cannot be used as a super-constructor in a subclass constructor.
That leaves no valid ways to refer to the constructor outside of the library.

(That's only generative construtors, factory constructors can be used on abstract classes.)
